src/config/cli.rs

```diff
@@ -18,6 +19,22 @@ pub(crate) struct Cli {
     #[clap(long)]
     pub(crate) etcd_endpoint: Option<String>,
 
+    // OpenShift supports encryption of:
+    // - Secrets, ConfigMaps encrypted by the kube-apiserver
+    // - Routes encrypted by the openshift-kube-apiserver
+    // - oauth {access,authorize} tokens encrypted by the openshift-oauth-apiserver
+    // The secrets/openshift-kube-apiserver/encryption-config and the secrets/openshift-oauth-apiserver/encryption-config
+    // are encrypted by the kube-apiserver and we need to decrypt those first in order to decrypt Routes and oauth token resources.
+    // Thus we resort to the kube-apiserver encryption-config file (i.e. /etc/kubernetes/static-pod-resources/kube-apiserver-pod-<latest>/secrets/encryption-config/encryption-config)
+    // to fetch the encryption details for the kube-apiserver encrypted resources (i.e. Secrets and ConfigMaps) first.
+    /// Kubernetes EncryptionConfiguration resource in JSON formatted string or path to the
+    /// respective file under /etc/kubernetes/static-pod-resources/kube-apiserver-pod-<latest>/secrets/encryption-config/encryption-config.
+    /// When specified, recert can be executed on OpenShift clusters with encryption at rest enabled.
+    /// At the end of its execution it will generate new encryption keys and use them to encrypt the
```

Comment:
Do we always want to re-encrypt? Probably yes, we never want the seed encryption keys. But do we always want to re-encrypt with random keys? Not so sure. What if some customers care about the particular keys with which their etcd database is encrypted? In that case, we would have to preserve those keys from the cluster being upgraded into the new cluster, so the new encryption keys should be configurable.

What do you think?

Reply:
Valid point! The [OpenShift etcd encryption docs](https://docs.openshift.com/container-platform/4.16/security/encrypting-etcd.html) do not address anything regarding the encryption keys and nor do the [OpenShift etcd backup ones](https://docs.openshift.com/container-platform/4.16/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html). When taking an etcd backup, alonside the etcd db there is a tarball with the `static-pod-resources` folder (which includes the `encryption-config` at that point in time).

```sh
$ ls -l
total 103176
-rw-------. 1 root root 105570336 Sep 23 12:57 snapshot_2024-09-23_125752.db
-rw-------. 1 root root     75121 Sep 23 12:57 static_kuberesources_2024-09-23_125752.tar.gz
```

I don't think users care about whether their encryption keys change or not. But I do see the point in implementing this feature like any such IBU feature, i.e. to keep the target cluster's encryption keys intact and add a boolean cmd line argument to forcefully rotate them when that's set (e.g. `--rotate-encryption-keys`).

That said, I would:

1. add a `--seed-encryption-config` cmd line arg and use that to decrypt
2. keep the current one, i.e. `--encryption-config` and use that one to encrypt and put it in the filesystem and etcd
3. implement the forceful encryption keys rotation in a subsequent PR

WDYT?

Reply:
What I suggested above is not going to be enough, as we need 3 encryption configs (`kube-apiserver`, `openshift-apiserver`, `oauth-apiserver`), so let's scratch that. I'll come up with a couple of alternatives.

Reply:
What I came up with (and started working on) is the following:

- auto-discover if seed encryption is enabled and the encryption type used (by reading the `apiserver/cluster`'s  `spec.encryption.type` field), then load the seed `encryption-config`s from the filesystem and etcd
- add 3 optional recert config args -> `{kube,openshift,oauth}-encryption-config`. When provided recert will use those to encrypt all resources to be encrypted, otherwise it will generate new `encryption-config`s
- put the `encryption-config`s used (either the user provided or the newly generated ones) in etcd and the filesystem

The above assumes that recert will  try to encrypt only when the seed is encrypted. In the case of the seed not being encrypted, recert will reject the `encryption-config` args. Does this make sense?